abi/abi-0.2.ts

```diff
@@ -0,0 +1,171 @@
+/// ABIs
+
+export interface AbiDefs {
+  functions?: FunctionDef[];
+  objects?: ObjectDef[];
+  enums?: EnumDef[];
+}
+
+export interface Abi extends AbiDefs {
+  version: "0.2";
+  imports?: ImportedAbi[];
+}
+
+export type ImportAbiType =
+  | "wasm"
+  | "interface";
+
+export interface ImportedAbi extends AbiDefs {
+  id: string;
+  uri: string;
+  type: ImportAbiType;
+  namespace: string;
+  imports?: ImportedAbi[];
+}
+
+/// Definitions (user-defined)
+
+export type UniqueDefKind =
+  | "Function"
+  | "Object"
+  | "Enum"
+
+export type DefKind =
+  | UniqueDefKind
+  | "Argument"
+  | "Result"
+  | "Property";
+
+export interface Def {
+  kind: DefKind;
+}
+
+export interface NamedDef extends Def {
+  name: string;
+}
+
+export interface InlinedTypeDef extends Def, OptionalType { }
+
+export interface NamedTypeDef extends NamedDef, InlinedTypeDef { }
+
+export interface FunctionDef extends NamedDef {
+  kind: "Function";
+  args: ArgumentDef[];
+  result: ResultDef;
+}
+
+export interface ArgumentDef extends NamedTypeDef {
+  kind: "Argument";
+}
+
+export interface ResultDef extends InlinedTypeDef {
+  kind: "Result";
+}
+
+export interface ObjectDef extends NamedDef {
+  kind: "Object";
+  props: PropertyDef[];
+}
+
+export interface PropertyDef extends NamedTypeDef {
+  kind: "Property";
+}
+
+export interface EnumDef extends NamedDef {
+  kind: "Enum";
+  constants: string[];
+}
+
+/// Types (built-ins)
+
+export type AnyType =
+  | ScalarType
+  | ArrayType
+  | MapType
```

Comment:
MapType should be removed and we should add ExternalType

Reply:
Same as above on custom types being re-worked separately; but acknowledged